Adobe Commerce (Magento) to Salesforce - Products and Prices

FrendsSarlinS opened the pull request:
#78 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added a new template for integrating Magento products and prices into Salesforce, including detailed implementation and usage notes.
  - Introduced metadata support for Adobe Commerce/Magento products integration with Salesforce.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jannejjj commented on Templates/CRMs/Magento/Magento Products to Salesforce/long-description.md:
```diff
@@ -0,0 +1,21 @@
+
+# Implementation and Usage Notes
+
+This template uses Salesforce queries to identify a product and the price book it's connected to. Price book, in which the product price is inserted, can be changed by altering the price book task parameters.
```
This sentence could probably be something like "The name of the price book in which the product prices are inserted, can be changed .... "

FrendsSarlinS replied:
Technically we change the whole price book, not just the name, but this gave me an idea 😄 "The price book, in which the product price is inserted, can be changed by altering the name in the price book task parameters."

jannejjj commented on Templates/CRMs/Magento/Magento Products to Salesforce/long-description.md:
```diff
@@ -0,0 +1,21 @@
+This template assumes that the following prerequisites are in place:
+
+- The Salesforce user should have access to client ID, secret and security token.
+- Price book for Magento is initialized in Salesforce.
```
Maybe we should mention that the name of said price book should be "Magento Price Book" by default 🤔

FrendsSarlinS replied:
This will be in the next commit!

jannejjj reviewed (commented):
Other than the comments i left directly on the files, I think in the step "Insert products to Salesforce" could be renamed to just "Insert product to Salesforce" (so change plural to singular), since we're only inserting one product per iteration in the for-loop. Additionally, it might make sense to have a text annotation connected to that task saying something like "When inserted, products are connected to the standard price book without a unit price" -> then we're adding the unit price information in the later steps. :D

FrendsSarlinS commented:
New commit is in, thanks Janne for the comments!

jannejjj commented on Templates/CRMs/Magento/Magento Products to Salesforce/process.json:
```diff
@@ -0,0 +1,85 @@
+			"Description": "This process will retrieve sales products from Adobe Commerce (Magento). It will perform the necessary mapping to Salesforce sales product format. Then it will transfer the mapped sales products to Salesforce.",
+			"Version": 1,
+			"UniqueIdentifier": "536e22f6-70cb-489c-9504-12852f515ea6",
+			"ProcessVariablesJson": "{\"MagentoBaseUrl\":{\"Value\":\"\\\"http://192.168.0.1/\\\"\",\"IsSecret\":false,\"Mode\":\"text\",\"Description\":\"Base Url for Magento.\"},\"MagentoUser\":{\"Value\":\"\\\"MagentoUsername\\\"\",\"IsSecret\":false,\"Mode\":\"text\",\"Description\":\"Username for Magento.\"},\"MagentoPassword\":{\"Value\":\"\\\"Password\\\"\",\"IsSecret\":true,\"Mode\":\"text\",\"Des ...
```
Thanks @coderabbitai, but those credentials are placeholders, not sensitive information.

jannejjj reviewed (approved):
After the latest changes, everything looks good and the template imports products into Salesforce correctly! As a result, the imported products can be used by the template that imports orders into Salesforce.

jefim commented on Templates/CRMs/Magento/Magento Products to Salesforce/assets/Adobe_Commerce___Magento_Product_to_Salesforce.svg:
1. In Magento token task turn off logging of parameters and results
2. In Magento token was not obtained - include Magento error (if it gives us anything meaningful) and at least include the http status code in the throw shape. Same goes for the throw error shape after getting product info.
3. Transform products and prices to JToken task seems redundant as HTTP task already return JToken
4. When doing error after "Retrieving product information successful?" also add more detailed on the error from Salesforce.
5. In `Get products from Salesforce` should we search with product code instead of product name? Also should the shape be called "Try getting product from Salesforce" or something similar?
6. Just in general - we should go over all those error message generation and include more info in them, such as e.g. Salesforce ErrorMessage.
7. When constructing messages you have several `JObject.FromObject` called on Salesforce results. But isnt Salesforce result already a JToken?

FrendsSarlinS replied:
Changes have now been implemented in the new commit!

jefim commented on Templates/CRMs/Magento/Magento Products to Salesforce/long-description.md:
```diff
@@ -0,0 +1,21 @@
+
+# Implementation and Usage Notes
+
+This template uses Salesforce queries to identify a product and the price book it's connected to. The price book, in which the product price is inserted, can be changed by altering the name in the price book task parameters.
```
Should this template handle price updates as well? As a template user I would probably see this as a major flaw, if those prices are only moved to Salesforce once and then would never get updated. Is this a hard thing to do?

FrendsSarlinS replied:
I think it could be implemented after retrieving Magento price book information: We try to get existing price book entry information that matches with the product code. If no lines are returned, we insert the product as a new price book entry. If an entry is returned, we update the data with new price information that's found in Magento.

Do you want me to give it a go?

jefim commented on Templates/CRMs/Magento/Magento Products to Salesforce/long-description.md:
```diff
@@ -0,0 +1,21 @@
+# Implementation and Usage Notes
+
+This template uses Salesforce queries to identify a product and the price book it's connected to. The price book, in which the product price is inserted, can be changed by altering the name in the price book task parameters.
+Process variables include fields for both Salesforce and Magento credentials.
```
Also, can you elaborate here the logic of searching for that correct pricebook? I mean the search for standard pricebook and checking if it is magento one and all that.

FrendsSarlinS replied:
Logic will be opened more in the next commit. Thanks!

jefim commented on Templates/CRMs/Magento/Magento Products to Salesforce/long-description.md:
```diff
@@ -0,0 +1,21 @@
+
+# Error Handling
+
+Error handling is not implemented in this template.
```
The error handling is quite extensive in the template. Please reflect it here (briefly). You can state that transient errors are not handled though.

FrendsSarlinS replied:
Good point, thanks! This will be taken into account in future commit.

jefim reviewed (commented):
The magento system should be under a new /Templates/E-commerce folder

jannejjj commented:
I noticed that the throw after getting product and price info from Magento says "An exception occurred when trying to get customer data .. ", when it should say product data 😄

FrendsSarlinS commented:
> I noticed that the throw after getting product and price info from Magento says "An exception occurred when trying to get customer data .. ", when it should say product data 😄

This has now been corrected, I also added descriptions to returns inside the loop and changed throws from expressions to text.

FrendsSarlinS commented:
Description and metadata files are now edited accordingly!
